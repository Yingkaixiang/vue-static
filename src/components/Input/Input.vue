<template>
  <textarea
    v-if="isTextArea"
    v-model="inputValue"
    ref="textarea"
    class="ct-input"
    :placeholder="placeholder"
    :rows="rows"
    @input="onInput"
    @focus="onFocus"
    @blur="onBlur"
  />
  <input
    v-else
    v-model="inputValue"
    class="ct-input"
    :placeholder="placeholder"
    @input="onInput"
    @focus="onFocus"
    @blur="onBlur"
  />
</template>

<script>
import { removeHTML, trim } from "../util/";

export default {
  model: {
    prop: "value",
    event: "input",
  },
  props: {
    value: [String, Number],
    type: String,
    placeholder: String,
    rows: {
      type: [String, Number],
      default: 1,
    },
  },
  data() {
    return {
      inputValue: this.format(this.value),
      isTextArea: this.type === "textarea",
    };
  },
  watch: {
    value(newVal) {
      this.inputValue = this.format(newVal);
      if (this.isTextArea) {
        this.$nextTick(this.adjustSize);
      }
    },
  },
  mounted() {
    if (this.isTextArea) {
      this.$nextTick(this.adjustSize);
    }
  },
  methods: {
    onInput() {
      this.$emit("input", this.inputValue);
      this.$parent.$emit("form.change", [this.inputValue]);
    },
    onFocus(e) {
      this.$emit("focus", e);
    },
    onBlur(e) {
      this.$emit("input", this.inputValue);
      this.$emit("blur", e);
      this.$parent.$emit("form.blur", [this.inputValue]);
    },
    format(value) {
      return this.type === "textarea"
        ? removeHTML(value)
        : removeHTML(trim(value));
    },
    adjustSize() {
      const { textarea } = this.$refs;
      textarea.style.height = "auto";
      let height = textarea.scrollHeight;
      textarea.style.height = `${height}px`;
    },
  },
};
</script>

<style lang="less">
.ct-input {
  width: 100%;
  border: 0;
  outline: none;
  background: transparent;
  resize: none;

  &:focus {
    outline: none;
  }
}
</style>
