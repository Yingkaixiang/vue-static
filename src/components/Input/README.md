# Input

自带 XSS 攻击防护。

## Attributes

| 参数 | 说明 | 类型 | 默认值 |
| - | - | - | - |
| placeholder | 输入框提示文案 | String | - |
| rows | textarea 行数 | String \| Number | 1 |
| type | 输入框类型 input 或 textarea | String | 'input' |
| value/v-model | 绑定值 | String \| Number | - |

## Events

| 参数 | 说明 | 回调参数 |
| - | - | - |
| blur | Input 失去焦点时触发 | (e: Event) |
| focus | Input 获取焦点时触发 | (e: Event) |
| input | Input 值改变时触发 | (value: String \| Number) |
