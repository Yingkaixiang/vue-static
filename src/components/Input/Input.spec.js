import { mount } from "@vue/test-utils";
import Input from "./Input";

describe("Input.vue", () => {
  it("Input Attributes", () => {
    const inputPlaceholder = "请输入姓名";
    const inputValue = "趣头条";

    const input = mount(Input, {
      propsData: {
        placeholder: inputPlaceholder,
        value: inputValue,
      },
    });

    const attr = input.attributes();
    expect(input.find("input").is("textarea")).toBe(false);
    expect(attr.placeholder).toBe(inputPlaceholder);
    expect(input.element).toMatchSnapshot();
  });

  it("Input Events", () => {
    const fnInput = jest.fn();
    const fnFocus = jest.fn();
    const fnBlur = jest.fn();

    const wrapper = mount(Input, {
      listeners: {
        input: fnInput,
        focus: fnFocus,
        blur: fnBlur,
      },
    });
    const input = wrapper.find("input");

    input.element.value = "1";
    input.trigger("input");
    expect(fnInput).toHaveBeenCalled();
    expect(wrapper.emitted("input")[0][0]).toEqual("1");

    input.trigger("focus");
    expect(fnFocus).toHaveBeenCalled();

    input.trigger("blur");
    expect(fnBlur).toHaveBeenCalled();
  });

  it("TextArea Attributes", () => {
    const textareaPlaceholder = "请输入姓名";
    const textareaValue = "趣头条";
    const textareaRows = "3";

    const textarea = mount(Input, {
      propsData: {
        type: "textarea",
        placeholder: textareaPlaceholder,
        value: textareaValue,
        rows: textareaRows,
      },
    });

    const attr = textarea.attributes();
    expect(textarea.find("textarea").is("textarea")).toBe(true);
    expect(attr.placeholder).toBe(textareaPlaceholder);
    expect(attr.rows).toBe(textareaRows);
    expect(textarea.element).toMatchSnapshot();
  });

  it("TextArea Events", () => {
    const textareaOnInput = jest.fn();
    const textareaOnFocus = jest.fn();
    const textareaOnBlur = jest.fn();

    const input = mount(Input, {
      listeners: {
        input: textareaOnInput,
        focus: textareaOnFocus,
        blur: textareaOnBlur,
      },
    });

    input.trigger("input");
    input.trigger("focus");
    input.trigger("blur");
    expect(textareaOnInput).toHaveBeenCalled();
    expect(textareaOnFocus).toHaveBeenCalled();
    expect(textareaOnBlur).toHaveBeenCalled();
  });

  it("Input value with HTML and space", () => {
    const inputValue = "   <script>console.log()</script> ";

    const input = mount(Input, {
      propsData: {
        value: inputValue,
      },
    });

    expect(input.element.value).toBe("console.log()");
  });

  it("TextArea value with HTML and space", () => {
    const inputValue = "  <script>console.log()</script> ";

    const input = mount(Input, {
      propsData: {
        type: "textarea",
        value: inputValue,
      },
    });

    expect(input.element.value).toBe("  console.log() ");
  });
});
