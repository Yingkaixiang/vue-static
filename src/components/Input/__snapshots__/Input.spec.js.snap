// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input.vue Input Attributes 1`] = `
<input
  class="ct-input"
  placeholder="请输入姓名"
/>
`;

exports[`Input.vue TextArea Attributes 1`] = `
<textarea
  class="ct-input"
  placeholder="请输入姓名"
  rows="3"
/>
`;
